### PR TITLE
PlaceInfo use case - Nicolas

### PR DESCRIPTION
**Pull request by nicolasdiasm**
[Nov 27: PlaceInfo use case implemented, back button added to SearchView and Clean Architecture fixes.

New files added: ListingUeCaseFactory, FilePlaceInfoDataAccessObject, PlaceInfoController, PlaceInfoPresenter, PlaceInfoState, PlaceInfoViewModel, PlaceInfoView, PlaceInfoDataAccessInterface, PlaceInfoInteractor, PlaceInfoInputData, PlaceInfoOutputData, PlaceInfoInputBoundary, PlaceInfoOutputBoundary.

Use Case Description: After getting a listing from the search use case, the user can now click on any place and retrieve more information about that specific place (name, address, city, coordinates, tags, and a google maps link).

### CODE REVIEW

**Comment by Arnnav-A** [comment]
We should wait on merging this, after review, until we merge Andrew's PR as the conflicts would be different.

File: src/interface_adapter/listing_results/ListingResultsState.java [hunk 1]
```diff
@@ -2,17 +2,24 @@
...
 import entity.Listing;
 
+import java.util.ArrayList;
+
 public class ListingResultsState {
-    private Listing listing;
+    private ArrayList<String> places_names;
     public String city;
     public String filter;
+    private String place_name;
```
Comment (JG808) [inline src/interface_adapter/listing_results/ListingResultsState.java:11]: String place_names is not a Java convention name, it's better to refactor

File: src/interface_adapter/listing_results/ListingResultsState.java [hunk 2]
```diff
@@ -23,8 +30,12 @@ public String getFilter() {
         return filter;
     }
 
-    public void setListing(Listing listing) {
-        this.listing = listing;
+    public String getPlaceName() {
```
Comment (JG808) [inline src/interface_adapter/listing_results/ListingResultsState.java:33]: add a line between setListing and getPlaceName

File: src/interface_adapter/listing_results/ListingResultsViewModel.java [hunk 3]
```diff
@@ -30,8 +32,8 @@ public String getFilter() {
         return filter;
     }
 
-    public ArrayList<Place> getPoints() {
-        return points;
+    public ArrayList<String> getPlacesName() {
```
Comment (JG808) [inline src/interface_adapter/listing_results/ListingResultsViewModel.java:35]: add a line between getPoints and getPlacesName

File: src/interface_adapter/listing_results/ListingResultsViewModel.java [hunk 4]
```diff
@@ -46,8 +48,8 @@ public void setFilter(String filter) {
         this.filter = filter;
     }
 
-    public void setPoints(ArrayList<Place> points) {
-        this.points = points;
+    public void setPlacesName(ArrayList<String> placesName) {
```
Comment (JG808) [inline src/interface_adapter/listing_results/ListingResultsViewModel.java:51]: add a line between setPoints and setPlacesName

File: src/interface_adapter/place_info/PlaceInfoPresenter.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package interface_adapter.place_info;
+
+import interface_adapter.ViewManagerModel;
+import use_case.place_info.PlaceInfoOutputBoundary;
+import use_case.place_info.PlaceInfoOutputData;
+import view.PlaceInfoView;
+import view.ViewManager;
+
```
Comment (JG808) [inline src/interface_adapter/place_info/PlaceInfoPresenter.java:8]: delete the line

File: src/interface_adapter/place_info/PlaceInfoState.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package interface_adapter.place_info;
+
+public class PlaceInfoState {
+
+    private String place_name;
```
Comment (JG808) [inline src/interface_adapter/place_info/PlaceInfoState.java:5]: Refacor place_name to a Java convention

File: src/interface_adapter/place_info/PlaceInfoViewModel.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package interface_adapter.place_info;
+
+import interface_adapter.ViewModel;
+
```
Comment (JG808) [inline src/interface_adapter/place_info/PlaceInfoViewModel.java:4]: delete the line

File: src/interface_adapter/listing_results/ListingResultsViewModel.java [hunk 8]
```diff
@@ -46,8 +48,8 @@ public void setFilter(String filter) {
...
+        this.placesName = placesName;
     }
 
     private final PropertyChangeSupport support = new PropertyChangeSupport(this);
```
Comment (JG808) [inline src/interface_adapter/listing_results/ListingResultsViewModel.java:55]: move it to the front

File: src/use_case/search/SearchInputData.java [hunk 9]
```diff
@@ -5,7 +5,6 @@ public class SearchInputData {
     private final String city;
     private final String filter;
 
-
```
Comment (JG808) [inline src/use_case/search/SearchInputData.java:8]: delet the "-"

File: src/use_case/search/SearchOutputData.java [hunk 10]
```diff
@@ -2,26 +2,24 @@
...
 
-public class SearchOutputData {
+import java.util.ArrayList;
 
-    private final Listing listing;
+public class SearchOutputData {
 
+    private final ArrayList<String> places_names;
     private final String city;
-
```
Comment (JG808) [inline src/use_case/search/SearchOutputData.java:10]: delete the line

File: src/use_case/search/SearchOutputData.java [hunk 11]
```diff
@@ -2,26 +2,24 @@
 
 import entity.Listing;
 
-public class SearchOutputData {
+import java.util.ArrayList;
 
-    private final Listing listing;
+public class SearchOutputData {
 
+    private final ArrayList<String> places_names;
```
Comment (JG808) [inline src/use_case/search/SearchOutputData.java:9]: Refactor places_names to Java convention

File: src/view/ListingView.java [hunk 12]
```diff
@@ -1,33 +1,40 @@
 package view;
 
-import entity.CommonListing;
-import entity.Place;
+import interface_adapter.ViewManagerModel;
 import interface_adapter.listing_results.ListingResultsState;
 import interface_adapter.listing_results.ListingResultsViewModel;
+import interface_adapter.place_info.PlaceInfoController;
```
Comment (JG808) [inline src/view/ListingView.java:7]: delete the line

File: src/view/PlaceInfoView.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
+package view;
+
+import interface_adapter.ViewManagerModel;
+import interface_adapter.place_info.PlaceInfoState;
+import interface_adapter.place_info.PlaceInfoViewModel;
+
```
Comment (JG808) [inline src/view/PlaceInfoView.java:6]: delete the line

**Review by JG808 — DISMISSED** [review]
Overall good, just some deletion of lines in import and name convention

**Comment by nicolasdiasm** [comment]
Finished changes

File: src/interface_adapter/place_info/PlaceInfoPresenter.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    @Override
+    public void prepareView(PlaceInfoOutputData placeInfoOutputData) {
+        PlaceInfoState placeInfoState = this.placeInfoViewModel.getState();
+        placeInfoState.setPlaceName(placeInfoOutputData.getName());
+        placeInfoState.setAddress(placeInfoOutputData.getAddress());
+        placeInfoState.setCoordinates(placeInfoOutputData.getCoordinates());
+        placeInfoState.setTags(placeInfoOutputData.getTags());
+        placeInfoState.setCity(placeInfoOutputData.getCity());
+        this.placeInfoViewModel.setState(placeInfoState);
+        this.placeInfoViewModel.firePropertyChanged();
```
Comment (Andrewwf1) [inline src/interface_adapter/place_info/PlaceInfoPresenter.java:26]: How about "place details not found"? This message should not be presented to user via an info window as usual but should pop up a window (for example) to warn user that this place has no information, or just delete this place when grabbing info from api (as we supposed to recommend places to user)

**Review by Andrewwf1 — COMMENTED** [review]
I go through all the modifications and think all the implementations are good. However, when I testing the actual performance, I find the placeInfo displays empty or "place not found" message when the API cannot find this place or lack of information about it. I think this can be improved by adding an if-else statement in placeInfo Presenter to check whether the selected place has useful information to display or not. Or you can choose to delete them during Place Factory, to avoid showing these empty places to user.
image

**Review by Arnnav-A — APPROVED** [review]
Approved by both reviewers, bug found by Andrew has been fixed

**Pull request merged**